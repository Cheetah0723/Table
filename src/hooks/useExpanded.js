import { useMemo } from 'react'
import PropTypes from 'prop-types'

import { getBy, getFirstDefined, setBy } from '../utils'
import { addActions, actions } from '../actions'
import { defaultState } from './useTableState'

defaultState.expanded = {}

addActions({
  toggleExpanded: '__toggleExpanded__',
  useExpanded: '__useExpanded__'
})

const propTypes = {
  expandedKey: PropTypes.string,
  paginateSubRows: PropTypes.bool
}

export const useExpanded = props => {
  PropTypes.checkPropTypes(propTypes, props, 'property', 'useExpanded')

  const {
    debug,
    rows,
    expandedKey = 'expanded',
    hooks,
    state: [{ expanded }, setState],
    paginateSubRows = true
  } = props

  const toggleExpandedByPath = (path, set) => {
    return setState(old => {
      const { expanded } = old
      const existing = getBy(expanded, path)
      set = getFirstDefined(set, !existing)
      return {
        ...old,
        expanded: setBy(expanded, path, set)
      }
    }, actions.toggleExpanded)
  }

  hooks.row.push(row => {
    const { path } = row
    row.toggleExpanded = set => toggleExpandedByPath(path, set)
  })

  const expandedRows = useMemo(() => {
    if (debug) console.info('getExpandedRows')

    const expandedRows = []

    // Here we do some mutation, but it's the last stage in the
    // immutable process so this is safe
    const handleRow = (row, index, depth = 0, parentPath = []) => {
      // Compute some final state for the row
      const path = [...parentPath, index]

      row.path = path
      row.depth = depth

      row.isExpanded =
        (row.original && row.original[expandedKey]) || getBy(expanded, path)

      if (paginateSubRows || (!paginateSubRows && row.depth === 0)) {
        expandedRows.push(row)
      }

      if (row.isExpanded && row.subRows && row.subRows.length) {
        row.subRows.forEach((row, i) => handleRow(row, i, depth + 1, path))
      }

      return row
    }

    rows.forEach((row, i) => handleRow(row, i))

    return expandedRows
  }, [debug, rows, expandedKey, expanded, paginateSubRows])

  const expandedDepth = findExpandedDepth(expanded)

  return {
    ...props,
    toggleExpandedByPath,
    expandedDepth,
    rows: expandedRows
  }
}

function findExpandedDepth(obj, depth = 1) {
  return Object.values(obj).reduce((prev, curr) => {
    if (typeof curr === 'object') {
      return Math.max(prev, findExpandedDepth(curr, depth + 1))
    }
    return depth
  }, 0)
}
