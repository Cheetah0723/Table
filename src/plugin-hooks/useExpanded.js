import React from 'react'

import {
  actions,
  makePropGetter,
  expandRows,
  useMountedLayoutEffect,
  useGetLatest,
} from '../utils'
import { useConsumeHookGetter, functionalUpdate } from '../publicUtils'

// Actions
actions.toggleExpanded = 'toggleExpanded'
actions.toggleAllExpanded = 'toggleAllExpanded'
actions.setExpanded = 'setExpanded'
actions.resetExpanded = 'resetExpanded'

export const useExpanded = hooks => {
  hooks.getExpandedToggleProps = [defaultGetExpandedToggleProps]
  hooks.stateReducers.push(reducer)
  hooks.useInstance.push(useInstance)
}

useExpanded.pluginName = 'useExpanded'

const defaultGetExpandedToggleProps = (props, { row }) => [
  props,
  {
    onClick: e => {
      e.persist()
      row.toggleExpanded()
    },
    style: {
      cursor: 'pointer',
    },
    title: 'Toggle Expanded',
  },
]

// Reducer
function reducer(state, action, previousState, instance) {
  if (action.type === actions.init) {
    return {
      expanded: {},
      ...state,
    }
  }

  if (action.type === actions.resetExpanded) {
    return {
      ...state,
      expanded: instance.initialState.expanded || {},
    }
  }

  if (action.type === actions.setExpanded) {
    return {
      ...state,
      expanded: functionalUpdate(action.expanded, state.expanded),
    }
  }

  if (action.type === actions.toggleExpanded) {
    const { id, expanded: setExpanded } = action
    const exists = state.expanded[id]

    const shouldExist =
      typeof setExpanded !== 'undefined' ? setExpanded : !exists

    if (!exists && shouldExist) {
      return {
        ...state,
        expanded: {
          ...state.expanded,
          [id]: true,
        },
      }
    } else if (exists && !shouldExist) {
      const { [id]: _, ...rest } = state.expanded
      return {
        ...state,
        expanded: rest,
      }
    } else {
      return state
    }
  }
}

function useInstance(instance) {
  const {
    data,
    rows,
    manualExpandedKey = 'expanded',
    paginateExpandedRows = true,
    expandSubRows = true,
    hooks,
    autoResetExpanded = true,
    state: { expanded },
    dispatch,
  } = instance

  const getAutoResetExpanded = useGetLatest(autoResetExpanded)

  // Bypass any effects from firing when this changes
  useMountedLayoutEffect(() => {
    if (getAutoResetExpanded()) {
      dispatch({ type: actions.resetExpanded })
    }
  }, [dispatch, data])

  const toggleExpanded = (id, expanded) => {
    dispatch({ type: actions.toggleExpanded, id, expanded })
  }

  // use reference to avoid memory leak in #1608
  const getInstance = useGetLatest(instance)

  const getExpandedTogglePropsHooks = useConsumeHookGetter(
    getInstance().hooks,
    'getExpandedToggleProps'
  )

  hooks.prepareRow.push(row => {
    row.toggleExpanded = set => instance.toggleExpanded(row.id, set)

    row.getExpandedToggleProps = makePropGetter(getExpandedTogglePropsHooks(), {
      instance: getInstance(),
      row,
    })
  })

  const expandedRows = React.useMemo(() => {
    if (paginateExpandedRows) {
      return expandRows(rows, { manualExpandedKey, expanded, expandSubRows })
    }

    return rows
  }, [paginateExpandedRows, rows, manualExpandedKey, expanded, expandSubRows])

  const expandedDepth = React.useMemo(() => findExpandedDepth(expanded), [
    expanded,
  ])

  Object.assign(instance, {
    preExpandedRows: rows,
    expandedRows,
    rows: expandedRows,
    toggleExpanded,
    expandedDepth,
  })
}

function findExpandedDepth(expanded) {
  let maxDepth = 0

  Object.keys(expanded).forEach(id => {
    const splitId = id.split('.')
    maxDepth = Math.max(maxDepth, splitId.length)
  })

  return maxDepth
}
