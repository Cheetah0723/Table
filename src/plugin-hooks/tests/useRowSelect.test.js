import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { useTable } from '../../hooks/useTable'
import { useRowSelect } from '../useRowSelect'
import { useExpanded } from '../useExpanded'

const dataPiece = [
  {
    firstName: 'tanner',
    lastName: 'linsley',
    age: 29,
    visits: 100,
    status: 'In Relationship',
    progress: 50,
  },
  {
    firstName: 'derek',
    lastName: 'perkins',
    age: 40,
    visits: 40,
    status: 'Single',
    progress: 80,
  },
  {
    firstName: 'joe',
    lastName: 'bergevin',
    age: 45,
    visits: 20,
    status: 'Complicated',
    progress: 10,
    expanded: true,
    subRows: [
      {
        firstName: 'bob',
        lastName: 'ross',
        age: 52,
        visits: 40,
        status: 'In Relationship',
        progress: 30,
      },
      {
        firstName: 'john',
        lastName: 'smith',
        age: 21,
        visits: 30,
        status: 'Single',
        progress: 60,
      },
    ],
  },
  {
    firstName: 'jaylen',
    lastName: 'linsley',
    age: 26,
    visits: 99,
    status: 'In Relationship',
    progress: 70,
  },
]

const data = [
  ...dataPiece,
  ...dataPiece,
  ...dataPiece,
  ...dataPiece,
  ...dataPiece,
  ...dataPiece,
]

function Table({ columns, data }) {
  // Use the state and functions returned from useTable to build your UI
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    state: { selectedRowIds },
  } = useTable(
    {
      columns,
      data,
    },
    useRowSelect,
    useExpanded,
    hooks => {
      hooks.flatColumns.push(columns => [
        // Let's make a column for selection
        {
          id: 'selection',
          // The header can use the table's getToggleAllRowsSelectedProps method
          // to render a checkbox
          Header: ({ getToggleAllRowsSelectedProps }) => (
            <div>
              <label>
                <IndeterminateCheckbox {...getToggleAllRowsSelectedProps()} />{' '}
                Select All
              </label>
            </div>
          ),
          // The cell can use the individual row's getToggleRowSelectedProps method
          // to the render a checkbox
          Cell: ({ row }) => (
            <div>
              <label>
                <IndeterminateCheckbox {...row.getToggleRowSelectedProps()} />{' '}
                Select Row
              </label>
            </div>
          ),
        },
        ...columns,
      ])
    }
  )

  // Render the UI for your table
  return (
    <>
      <table {...getTableProps()}>
        <thead>
          {headerGroups.map(headerGroup => (
            <tr {...headerGroup.getHeaderGroupProps()}>
              {headerGroup.headers.map(column => (
                <th {...column.getHeaderProps()}>{column.render('Header')}</th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {rows.map(
            (row, i) =>
              prepareRow(row) || (
                <tr {...row.getRowProps()}>
                  {row.cells.map(cell => {
                    return (
                      <td {...cell.getCellProps()}>{cell.render('Cell')}</td>
                    )
                  })}
                </tr>
              )
          )}
        </tbody>
      </table>
      <p>Selected Rows: {Object.keys(selectedRowIds).length}</p>
      <pre>
        <code>
          {JSON.stringify({ selectedRowIds: selectedRowIds }, null, 2)}
        </code>
      </pre>
    </>
  )
}

const IndeterminateCheckbox = React.forwardRef(
  ({ indeterminate, ...rest }, ref) => {
    const defaultRef = React.useRef()
    const resolvedRef = ref || defaultRef

    React.useEffect(() => {
      resolvedRef.current.indeterminate = indeterminate
    }, [resolvedRef, indeterminate])

    return <input type="checkbox" ref={resolvedRef} {...rest} />
  }
)

function App() {
  const columns = React.useMemo(
    () => [
      {
        id: 'selectedStatus',
        Cell: ({ row }) => (
          <div>
            Row {row.id} {row.isSelected ? 'Selected' : 'Not Selected'}
          </div>
        ),
      },
      {
        Header: 'Name',
        columns: [
          {
            Header: 'First Name',
            accessor: 'firstName',
          },
          {
            Header: 'Last Name',
            accessor: 'lastName',
          },
        ],
      },
      {
        Header: 'Info',
        columns: [
          {
            Header: 'Age',
            accessor: 'age',
          },
          {
            Header: 'Visits',
            accessor: 'visits',
          },
          {
            Header: 'Status',
            accessor: 'status',
          },
          {
            Header: 'Profile Progress',
            accessor: 'progress',
          },
        ],
      },
    ],
    []
  )

  return <Table columns={columns} data={data} />
}

test('renders a table with selectable rows', () => {
  const { getByLabelText, getAllByLabelText, asFragment } = render(<App />)

  const fragment1 = asFragment()

  fireEvent.click(getByLabelText('Select All'))

  const fragment2 = asFragment()

  fireEvent.click(getByLabelText('Select All'))

  const fragment3 = asFragment()

  fireEvent.click(getAllByLabelText('Select Row')[0])
  fireEvent.click(getAllByLabelText('Select Row')[2])

  const fragment4 = asFragment()

  fireEvent.click(getAllByLabelText('Select Row')[2])

  const fragment5 = asFragment()

  fireEvent.click(getAllByLabelText('Select Row')[3])

  const fragment6 = asFragment()
  fireEvent.click(getAllByLabelText('Select Row')[4])

  const fragment7 = asFragment()

  fireEvent.click(getAllByLabelText('Select Row')[4])

  const fragment8 = asFragment()

  expect(fragment1).toMatchDiffSnapshot(fragment2)
  expect(fragment2).toMatchDiffSnapshot(fragment3)
  expect(fragment3).toMatchDiffSnapshot(fragment4)
  expect(fragment4).toMatchDiffSnapshot(fragment5)
  expect(fragment5).toMatchDiffSnapshot(fragment6)
  expect(fragment6).toMatchDiffSnapshot(fragment7)
  expect(fragment7).toMatchDiffSnapshot(fragment8)
})
