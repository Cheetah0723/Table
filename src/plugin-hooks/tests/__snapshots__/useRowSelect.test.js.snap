// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a table with selectable rows 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -117,11 +117,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 0 Not Selected
+             Row 0 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -173,11 +173,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 1 Not Selected
+             Row 1 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -229,11 +229,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2 Not Selected
+             Row 2 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -285,11 +285,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.0 Not Selected
+             Row 2.0 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -341,11 +341,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.1 Not Selected
+             Row 2.1 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -397,11 +397,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 3 Not Selected
+             Row 3 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -453,11 +453,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 4 Not Selected
+             Row 4 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -509,11 +509,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 5 Not Selected
+             Row 5 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -565,11 +565,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 6 Not Selected
+             Row 6 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -621,11 +621,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 6.0 Not Selected
+             Row 6.0 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -677,11 +677,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 6.1 Not Selected
+             Row 6.1 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -733,11 +733,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 7 Not Selected
+             Row 7 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -789,11 +789,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 8 Not Selected
+             Row 8 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -845,11 +845,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 9 Not Selected
+             Row 9 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -901,11 +901,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 10 Not Selected
+             Row 10 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -957,11 +957,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 10.0 Not Selected
+             Row 10.0 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1013,11 +1013,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 10.1 Not Selected
+             Row 10.1 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1069,11 +1069,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 11 Not Selected
+             Row 11 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1125,11 +1125,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 12 Not Selected
+             Row 12 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1181,11 +1181,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 13 Not Selected
+             Row 13 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1237,11 +1237,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 14 Not Selected
+             Row 14 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1293,11 +1293,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 14.0 Not Selected
+             Row 14.0 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1349,11 +1349,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 14.1 Not Selected
+             Row 14.1 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1405,11 +1405,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 15 Not Selected
+             Row 15 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1461,11 +1461,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 16 Not Selected
+             Row 16 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1517,11 +1517,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 17 Not Selected
+             Row 17 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1573,11 +1573,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 18 Not Selected
+             Row 18 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1629,11 +1629,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 18.0 Not Selected
+             Row 18.0 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1685,11 +1685,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 18.1 Not Selected
+             Row 18.1 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1741,11 +1741,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 19 Not Selected
+             Row 19 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1797,11 +1797,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 20 Not Selected
+             Row 20 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1853,11 +1853,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 21 Not Selected
+             Row 21 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1909,11 +1909,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 22 Not Selected
+             Row 22 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1965,11 +1965,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 22.0 Not Selected
+             Row 22.0 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2021,11 +2021,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 22.1 Not Selected
+             Row 22.1 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2077,11 +2077,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 23 Not Selected
+             Row 23 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2114,15 +2114,52 @@
          </td>
        </tr>
      </tbody>
    </table>
    <p>
-     Selected Rows: 0
+     Selected Rows: 36
    </p>
    <pre>
      <code>
        {
-   "selectedRowPaths": []
+   "selectedRowPaths": [
+     "0",
+     "1",
+     "2",
+     "2.0",
+     "2.1",
+     "3",
+     "4",
+     "5",
+     "6",
+     "6.0",
+     "6.1",
+     "7",
+     "8",
+     "9",
+     "10",
+     "10.0",
+     "10.1",
+     "11",
+     "12",
+     "13",
+     "14",
+     "14.0",
+     "14.1",
+     "15",
+     "16",
+     "17",
+     "18",
+     "18.0",
+     "18.1",
+     "19",
+     "20",
+     "21",
+     "22",
+     "22.0",
+     "22.1",
+     "23"
+   ]
  }
      </code>
    </pre>
  </DocumentFragment>
`;

exports[`renders a table with selectable rows 2`] = `
Snapshot Diff:
- First value
+ Second value

@@ -117,11 +117,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 0 Selected
+             Row 0 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -173,11 +173,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 1 Selected
+             Row 1 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -229,11 +229,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2 Selected
+             Row 2 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -285,11 +285,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.0 Selected
+             Row 2.0 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -341,11 +341,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.1 Selected
+             Row 2.1 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -397,11 +397,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 3 Selected
+             Row 3 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -453,11 +453,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 4 Selected
+             Row 4 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -509,11 +509,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 5 Selected
+             Row 5 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -565,11 +565,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 6 Selected
+             Row 6 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -621,11 +621,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 6.0 Selected
+             Row 6.0 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -677,11 +677,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 6.1 Selected
+             Row 6.1 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -733,11 +733,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 7 Selected
+             Row 7 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -789,11 +789,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 8 Selected
+             Row 8 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -845,11 +845,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 9 Selected
+             Row 9 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -901,11 +901,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 10 Selected
+             Row 10 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -957,11 +957,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 10.0 Selected
+             Row 10.0 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1013,11 +1013,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 10.1 Selected
+             Row 10.1 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1069,11 +1069,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 11 Selected
+             Row 11 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1125,11 +1125,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 12 Selected
+             Row 12 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1181,11 +1181,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 13 Selected
+             Row 13 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1237,11 +1237,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 14 Selected
+             Row 14 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1293,11 +1293,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 14.0 Selected
+             Row 14.0 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1349,11 +1349,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 14.1 Selected
+             Row 14.1 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1405,11 +1405,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 15 Selected
+             Row 15 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1461,11 +1461,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 16 Selected
+             Row 16 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1517,11 +1517,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 17 Selected
+             Row 17 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1573,11 +1573,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 18 Selected
+             Row 18 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1629,11 +1629,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 18.0 Selected
+             Row 18.0 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1685,11 +1685,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 18.1 Selected
+             Row 18.1 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1741,11 +1741,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 19 Selected
+             Row 19 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1797,11 +1797,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 20 Selected
+             Row 20 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1853,11 +1853,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 21 Selected
+             Row 21 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1909,11 +1909,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 22 Selected
+             Row 22 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -1965,11 +1965,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 22.0 Selected
+             Row 22.0 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2021,11 +2021,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 22.1 Selected
+             Row 22.1 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2077,11 +2077,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 23 Selected
+             Row 23 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2114,52 +2114,15 @@
          </td>
        </tr>
      </tbody>
    </table>
    <p>
-     Selected Rows: 36
+     Selected Rows: 0
    </p>
    <pre>
      <code>
        {
-   "selectedRowPaths": [
-     "0",
-     "1",
-     "2",
-     "2.0",
-     "2.1",
-     "3",
-     "4",
-     "5",
-     "6",
-     "6.0",
-     "6.1",
-     "7",
-     "8",
-     "9",
-     "10",
-     "10.0",
-     "10.1",
-     "11",
-     "12",
-     "13",
-     "14",
-     "14.0",
-     "14.1",
-     "15",
-     "16",
-     "17",
-     "18",
-     "18.0",
-     "18.1",
-     "19",
-     "20",
-     "21",
-     "22",
-     "22.0",
-     "22.1",
-     "23"
-   ]
+   "selectedRowPaths": []
  }
      </code>
    </pre>
  </DocumentFragment>
`;

exports[`renders a table with selectable rows 3`] = `
Snapshot Diff:
- First value
+ Second value

@@ -117,11 +117,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 0 Not Selected
+             Row 0 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -229,11 +229,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2 Not Selected
+             Row 2 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -285,11 +285,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.0 Not Selected
+             Row 2.0 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -341,11 +341,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.1 Not Selected
+             Row 2.1 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2114,15 +2114,20 @@
          </td>
        </tr>
      </tbody>
    </table>
    <p>
-     Selected Rows: 0
+     Selected Rows: 4
    </p>
    <pre>
      <code>
        {
-   "selectedRowPaths": []
+   "selectedRowPaths": [
+     "0",
+     "2",
+     "2.0",
+     "2.1"
+   ]
  }
      </code>
    </pre>
  </DocumentFragment>
`;

exports[`renders a table with selectable rows 4`] = `
Snapshot Diff:
- First value
+ Second value

@@ -229,11 +229,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2 Selected
+             Row 2 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -285,11 +285,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.0 Selected
+             Row 2.0 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -341,11 +341,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.1 Selected
+             Row 2.1 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2114,20 +2114,17 @@
          </td>
        </tr>
      </tbody>
    </table>
    <p>
-     Selected Rows: 4
+     Selected Rows: 1
    </p>
    <pre>
      <code>
        {
    "selectedRowPaths": [
-     "0",
-     "2",
-     "2.0",
-     "2.1"
+     "0"
    ]
  }
      </code>
    </pre>
  </DocumentFragment>
`;

exports[`renders a table with selectable rows 5`] = `
Snapshot Diff:
- First value
+ Second value

@@ -285,11 +285,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.0 Not Selected
+             Row 2.0 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2114,17 +2114,18 @@
          </td>
        </tr>
      </tbody>
    </table>
    <p>
-     Selected Rows: 1
+     Selected Rows: 2
    </p>
    <pre>
      <code>
        {
    "selectedRowPaths": [
-     "0"
+     "0",
+     "2.0"
    ]
  }
      </code>
    </pre>
  </DocumentFragment>
`;

exports[`renders a table with selectable rows 6`] = `
Snapshot Diff:
- First value
+ Second value

@@ -341,11 +341,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.1 Not Selected
+             Row 2.1 Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2114,18 +2114,19 @@
          </td>
        </tr>
      </tbody>
    </table>
    <p>
-     Selected Rows: 2
+     Selected Rows: 3
    </p>
    <pre>
      <code>
        {
    "selectedRowPaths": [
      "0",
-     "2.0"
+     "2.0",
+     "2.1"
    ]
  }
      </code>
    </pre>
  </DocumentFragment>
`;

exports[`renders a table with selectable rows 7`] = `
Snapshot Diff:
- First value
+ Second value

@@ -341,11 +341,11 @@
          </td>
          <td
            class=""
          >
            <div>
-             Row 2.1 Selected
+             Row 2.1 Not Selected
            </div>
          </td>
          <td
            class=""
          >
@@ -2114,19 +2114,18 @@
          </td>
        </tr>
      </tbody>
    </table>
    <p>
-     Selected Rows: 3
+     Selected Rows: 2
    </p>
    <pre>
      <code>
        {
    "selectedRowPaths": [
      "0",
-     "2.0",
-     "2.1"
+     "2.0"
    ]
  }
      </code>
    </pre>
  </DocumentFragment>
`;
